// @twilio-labs/serverless-api/dist/api.js
"use strict";
/** @module @twilio-labs/serverless-api */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.utils = exports.api = exports.fsHelpers = void 0;
const internals = __importStar(require("./api/index"));
const fs = __importStar(require("./utils/fs"));
const _utils = __importStar(require("./utils/index"));
exports.fsHelpers = fs;
exports.api = internals;
exports.utils = _utils;
var builds_1 = require("./api/builds");
Object.defineProperty(exports, "activateBuild", { enumerable: true, get: function () { return builds_1.activateBuild; } });
Object.defineProperty(exports, "triggerBuild", { enumerable: true, get: function () { return builds_1.triggerBuild; } });
Object.defineProperty(exports, "waitForSuccessfulBuild", { enumerable: true, get: function () { return builds_1.waitForSuccessfulBuild; } });
var dependencies_1 = require("./api/dependencies");
Object.defineProperty(exports, "getDependencies", { enumerable: true, get: function () { return dependencies_1.getDependencies; } });
var functions_1 = require("./api/functions");
Object.defineProperty(exports, "getOrCreateFunctionResources", { enumerable: true, get: function () { return functions_1.getOrCreateFunctionResources; } });
Object.defineProperty(exports, "uploadFunction", { enumerable: true, get: function () { return functions_1.uploadFunction; } });
var variables_1 = require("./api/variables");
Object.defineProperty(exports, "setEnvironmentVariables", { enumerable: true, get: function () { return variables_1.setEnvironmentVariables; } });
