// twilio-run/dist/config/env/env-import.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfigFromFlags = void 0;
const path_1 = __importDefault(require("path"));
const list_1 = require("../../commands/list");
const utils_1 = require("../../serverless-api/utils");
const global_1 = require("../global");
const utils_2 = require("../utils");
const mergeFlagsAndConfig_1 = require("../utils/mergeFlagsAndConfig");
function getConfigFromFlags(flags, externalCliOptions) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
        let cwd = flags.cwd ? path_1.default.resolve(flags.cwd) : process.cwd();
        flags.cwd = cwd;
        if (flags.production) {
            flags.environment = '';
        }
        const configFlags = global_1.readSpecializedConfig(cwd, flags.config, 'env', {
            username: flags.username ||
                (externalCliOptions && externalCliOptions.accountSid) ||
                undefined,
            environmentSuffix: flags.environment,
        });
        flags = mergeFlagsAndConfig_1.mergeFlagsAndConfig(configFlags, flags, list_1.cliInfo);
        cwd = flags.cwd || cwd;
        const { localEnv: envFileVars, envPath } = yield utils_2.readLocalEnvFile(flags);
        const { username, password } = yield utils_2.getCredentialsFromFlags(flags, envFileVars, externalCliOptions);
        const serviceSid = flags.serviceSid ||
            (yield utils_1.getFunctionServiceSid(cwd, flags.config, 'env', ((_a = flags.username) === null || _a === void 0 ? void 0 : _a.startsWith('AC')) ? flags.username
                : username.startsWith('AC')
                    ? username
                    : externalCliOptions === null || externalCliOptions === void 0 ? void 0 : externalCliOptions.accountSid));
        let serviceName = yield utils_2.getServiceNameFromFlags(flags);
        if (!flags.key) {
            throw new Error('Missing --key argument. Please provide a key for your environment variable.');
        }
        if (!flags.value) {
            throw new Error('Missing --value argument. Please provide a key for your environment variable.');
        }
        const env = utils_2.filterEnvVariablesForDeploy(envFileVars);
        return {
            cwd,
            username,
            password,
            serviceSid,
            serviceName,
            environment: flags.environment,
            region: flags.region,
            edge: flags.edge,
            env,
            append: false,
        };
    });
}
exports.getConfigFromFlags = getConfigFromFlags;
