// @twilio/runtime-handler/dist/dev-runtime/utils/debug.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRedactedObject = exports.copyObject = exports.allPropertiesRedactor = exports.generalRedactor = void 0;
const debug_1 = __importDefault(require("debug"));
const fast_redact_1 = __importDefault(require("fast-redact"));
function prefixAllEntriesWithWildcard(values) {
    const result = [];
    for (let val of values) {
        result.push(val);
        result.push(`*.${val}`);
    }
    return result;
}
exports.generalRedactor = fast_redact_1.default({
    paths: [
        'env.*',
        'pkgJson.*',
        ...prefixAllEntriesWithWildcard([
            'authToken',
            'apiSecret',
            'username',
            'password',
            'cookies',
            'AUTH_TOKEN',
            'API_SECRET',
            'TWILIO_AUTH_TOKEN',
            'TWILIO_API_SECRET',
        ]),
    ],
    serialize: false,
});
exports.allPropertiesRedactor = fast_redact_1.default({
    paths: ['*'],
    serialize: false,
});
function copyObject(obj) {
    return JSON.parse(JSON.stringify(obj));
}
exports.copyObject = copyObject;
function createRedactedObject(obj, redactor) {
    const copiedObject = copyObject(obj);
    return redactor(copiedObject);
}
exports.createRedactedObject = createRedactedObject;
debug_1.default.formatters.P = function protectedFormatterMultiline(v) {
    if (typeof v === 'object') {
        v = createRedactedObject(v, exports.generalRedactor);
    }
    return debug_1.default.formatters.O.bind(debug_1.default)(v);
};
debug_1.default.formatters.p = function protectedFormatterSameline(v) {
    if (typeof v === 'object') {
        v = createRedactedObject(v, exports.generalRedactor);
    }
    return debug_1.default.formatters.o.bind(debug_1.default)(v);
};
debug_1.default.formatters.R = function redactedFormatterMultiline(v) {
    if (typeof v === 'object') {
        v = createRedactedObject(v, exports.allPropertiesRedactor);
    }
    return debug_1.default.formatters.O.bind(debug_1.default)(v);
};
debug_1.default.formatters.r = function redactedFormatterSameline(v) {
    if (typeof v === 'object') {
        v = createRedactedObject(v, exports.allPropertiesRedactor);
    }
    return debug_1.default.formatters.o.bind(debug_1.default)(v);
};
exports.default = debug_1.default;
