// @twilio/runtime-handler/dist/dev-runtime/internal/runtime.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = void 0;
const fs_1 = require("fs");
const check_account_sid_1 = require("../checks/check-account-sid");
const debug_1 = __importDefault(require("../utils/debug"));
const requireFromProject_1 = require("../utils/requireFromProject");
const route_cache_1 = require("./route-cache");
const log = debug_1.default('twilio-runtime-handler:dev:runtime');
function getAssets() {
    const { assets } = route_cache_1.getCachedResources();
    if (assets.length === 0) {
        return {};
    }
    const result = {};
    for (const asset of assets) {
        if (asset.access === 'private') {
            const prefix = process.env.TWILIO_FUNCTIONS_LEGACY_MODE === 'true' ? '/assets' : '';
            const open = () => fs_1.readFileSync(asset.filePath, 'utf8');
            result[prefix + asset.path] = { path: asset.filePath, open };
        }
    }
    log('Found the following assets available: %o', result);
    return result;
}
function getFunctions() {
    const { functions } = route_cache_1.getCachedResources();
    if (functions.length === 0) {
        return {};
    }
    const result = {};
    for (const fn of functions) {
        result[fn.path.substr(1)] = { path: fn.filePath };
    }
    log('Found the following functions available: %o', result);
    return result;
}
function create({ env, logger, baseDir, }) {
    function getSync(options) {
        options = Object.assign({ serviceName: 'default', lazyLoading: true }, options);
        const { serviceName } = options;
        delete options.serviceName;
        check_account_sid_1.checkForValidAccountSid(env.ACCOUNT_SID, {
            shouldPrintMessage: true,
            shouldThrowError: true,
            logger: logger,
            functionName: `Runtime.getSync(${[...arguments]
                .map((x) => JSON.stringify(x))
                .join(',')})`,
        });
        const client = requireFromProject_1.requireFromProject(baseDir, 'twilio', true)(env.ACCOUNT_SID, env.AUTH_TOKEN, options);
        const service = client.sync.services(serviceName || 'default');
        service.maps = service.syncMaps;
        service.lists = service.syncLists;
        return service;
    }
    return { getSync, getAssets, getFunctions };
}
exports.create = create;
module.exports = { create };
